content/blog/eng/common-problems-faced-by-php-developers-in-consuming-an-ampq-message.md

```diff
@@ -0,0 +1,298 @@
+---
+authors: ["emulator000"]
+date: "2018-11-15"
+draft: true
+share: true
+categories: [English, RabbitMQ, AMQP, Consumer, PHP, Rust, Go]
+title: "Common problems faced by PHP developers in consuming an AMQP message"
+type: "post"
+languageCode: "en-EN"
+toc: true
+---
+
+When we use RabbitMQ and our project is in PHP, we have to run a Consumer and we could encounter some common and still unresolved problems.
+
+In order to implement RabbitMQ in the project, there are different ready to use packages, especially when we use a framework like Symfony or Laravel. Usually these packages provides the full needed integration for the AMQP protocol so, we can easily configure it and create a Producer, a Consumer and then, we have to create a little script that will supervisor the PHP process in order to make sure that all our consumers are running. 
+
+# Consumer
+
+In order to consume a message, we have to run a demon that will listen for new messages and then, process each message executing some logic; this is a normal flow in case of blocking requests, we don't want to block the user request nor loose it.
+
+We want also to keep track of all requests and if one fails, we could retry to process again or just ignore it if some blocking error occurs.
+
+## The memory problem
+
+A normal PHP script serves a single request, then, the script "dies" naturally as it ended the execution of provided instructions and frees the allocated memory. In a PHP consumer, we have to run a long running PHP process so we have to face the memory problem, yes, because as PHP is not properly designed to achieve this scope, this can cause troubles for sure if we don't pay attention to the script's code.
+
+Normally, you don't notice any memory issue due to fact that the script runs for few seconds, also, in a long running PHP proccess, if you try a simple script and make it run for few minutes, maybe they will not use any additional memory at all but if you use some framework, an ORM or you are in a complex application with many dependencies, the application will eventually allocate some memory (eg. for caching results) and you can notice that the PHP process will increase their memory hour by hour due to some memory leaks.
+
+That's not a problem at all in a normal case, the script ends and the memory is freed normally, this is something that is not so obvious in a long running process like how it would be a PHP consumer. They will break, leak and in some cases, corrupts the memory and this will make the process crash or more drastically, they will fail to run some instruction as the opcode cache gets changed while the process is running.
```

Comment:
```suggestion
That's not a problem at all in a normal case, where the script ends and the memory is freed normally, but the case is not that obvious for a long running process like the one required by a PHP consumer. It will break, leak and, in some cases, corrupt the memory, and this will make the process crash or, more drastically, fail to run some instructions, as the opcode cache gets changed while the process is running.
```